Update the How To doc of the Bindgen tool (#689)
ballerina-platform/ballerina-dev-website: learn/calling-java-code-from-ballerina.md

### Step 3: Generate Ballerina bindings 
In this step, we'll use the `bindgen` tool to generate Ballerina bindings for those four classes that we talked about in Step 1. If you want more information about the tool, you can refer [The `bindgen` tool](#the-bindgen-tool).

#### Generate Ballerina bindings
```sh
> ballerina bindgen -mvn org.yaml:snakeyaml:1.25 -o src/yamlparser org.yaml.snakeyaml.Yaml
 java.io.FileInputStream java.io.InputStream java.util.Map

Ballerina project detected at: /Users/sameera/yaml-project

Resolving maven dependencies...
snakeyaml-1.25.pom 100% [=================================================] 37/37 KB (0:00:00 / 0:00:00)
snakeyaml-1.25.jar 100% [===============================================] 297/297 KB (0:00:01 / 0:00:00)

> Updated the Ballerina.toml file with new platform libraries.

Reviewer: ```suggestion
Updated the `Ballerina.toml` file with the new platform libraries.
```

> Following jars were added to the classpath:

Reviewer: ```suggestion
The following JARs were added to the classpath:
```

	snakeyaml-1.25.jar

Generating bindings for:
	java.util.Map
	java.io.FileInputStream
	org.yaml.snakeyaml.Yaml
	java.io.InputStream

Generating dependency bindings for:
	org.yaml.snakeyaml.introspector.BeanAccess
	java.util.function.BiFunction
	org.yaml.snakeyaml.DumperOptions$FlowStyle
	...
	...
```

> - The `-mvn` option specifies the maven dependency of the Java library required to generate bindings.

Reviewer: ```suggestion
- The `-mvn` option specifies the Maven dependency of the Java library required to generate bindings.
```

- The `-o` option specifies the output directory to which the generated bindings are stored. In this case, we instruct the tool to store bindings inside the `yamlparser` module. 
- The argument list specifies the Java class names. 

> The `bindgen` tool generate bindings for 

Reviewer: ```suggestion
The `bindgen` tool generate bindings for:
```

> - The specified Java classes.

Reviewer: ```suggestion
- the specified Java classes
```

> - The Java classes exposed in the public APIs of all the specified classes. 

Reviewer: ```suggestion
- the Java classes exposed in the public APIs of all the specified classes
```

}
```
### Create the SnakeYAML entry point

> The `org.yaml.snakeyaml.Yaml` Class is the entry point to the SnakeYAML API.  The corresponding generated Ballerina object is `Yaml`. The `newYaml1()` function is mapped to the default constructor of the Java class.

Reviewer: ```suggestion
The `org.yaml.snakeyaml.Yaml` class is the entry point to the SnakeYAML API.  The generated corresponding Ballerina object is `Yaml`. The `newYaml1()` function is mapped to the default constructor of the Java class.
```

```ballerina
Yaml yaml = newYaml1();
```
###  Loads the YAML Document

> We'll be using the `org.yaml.snakeyaml.Yaml.load(InputStream is)` method to get a Java Map instance from the given InputStream. 

Reviewer: ```suggestion
We'll be using the `org.yaml.snakeyaml.Yaml.load(InputStream is)` method to get a Java Map instance from the given `InputStream`. 
```

```sh
ballerina bindgen [(-cp|--classpath) <classpath>...]
                  [(-mvn|--maven) <groupId>:<artifactId>:<version>]
                  [(-o|--output) <output>]
                  (<class-name>...)
```

`(-cp|--classpath) <classpath>...`

> This optional parameter could be used to specify one or more comma-delimited classpaths for retrieving the required Java libraries needed by the bindgen tool execution. The classpath could be provided as comma-separated paths of JAR files or as comma-separated paths of directories containing all the relevant Java libraries. If the Ballerina bindings are to be generated from a standard Java library, from a library available inside the Ballerina SDK or from a platform library specified in the `Ballerina.toml`, then you need not specify the classpath explicitly.

Reviewer: ```suggestion
This optional parameter could be used to specify one or more comma-delimited classpaths for retrieving the required Java libraries needed by the bindgen tool execution. The classpath could be provided as comma-separated paths of JAR files or as comma-separated paths of directories containing all the relevant Java libraries. If the Ballerina bindings are to be generated from a standard Java library, from a library available inside the Ballerina SDK, or from a platform library specified in the `Ballerina.toml`, then you need not specify the classpath explicitly.
```

`(-mvn|--maven) <groupId>:<artifactId>:<version>`

> This optional parameter could be used to specify a maven dependency required for the generation of Ballerina bindings. Here, the specified library and its transitive dependencies will be resolved into `target/platform-libs` directory of the project. If the tool is not executed inside a project or if the output path does not point to a project, a directory structure `target/platform-libs` will be created in the output path to store the maven dependencies. The tool will also update the `Ballerina.toml` file with the platform libraries if the command is executed inside a Ballerina project. If Java classes from multiple dependencies are to be generated, it should be done in stages. 

Reviewer: ```suggestion
This optional parameter could be used to specify a Maven dependency required for the generation of Ballerina bindings. Here, the specified library and its transitive dependencies will be resolved into the `target/platform-libs` directory of the project. If the tool is not executed inside a project or if the output path does not point to a project, the `target/platform-libs` directory structure will be created in the output path to store the Maven dependencies. The tool will also update the `Ballerina.toml` file with the platform libraries if the command is executed inside a Ballerina project. If Java classes from multiple dependencies are to be generated, it should be done in stages. 
```

> When the tool is run, a `.bal` file will be created to represent each Java class. This would contain the respective Ballerina object along with the required Java interoperability mappings. These `.bal` files would reside inside sub directories representing the package structure. Apart from creating bindings for the specified Java classes, the command would also generate empty Ballerina objects for the dependent Java classes. A Java class would be considered dependent if it is used inside one of the Ballerina objects generated. A set of additional utility files will also be generated in order to support the auto-generated Ballerina bindings. The folder structure of the generated bindings will be as follows.

Reviewer: ```suggestion
When the tool is run, a `.bal` file will be created to represent each Java class. This would contain the respective Ballerina object along with the required Java interoperability mappings. These `.bal` files would reside inside sub directories representing the package structure. Apart from creating bindings for the specified Java classes, the command would also generate empty Ballerina objects for the dependent Java classes. A Java class would be considered dependent if it is used inside one of the  generated Ballerina objects. A set of additional utility files will also be generated in order to support the auto-generated Ballerina bindings. The folder structure of the generated bindings will be as follows.
```

> These interop functions take the fully-qualified Java method name as the function name. However, if there exists overloaded methods, a numeric suffix will be appended at the end.

Reviewer: ```suggestion
These interop functions take the fully-qualified Java method name as the function name. However, if overloaded methods exist, a numeric suffix will be appended at the end.
```

};
```

#### Type mappings between Java and Ballerina
In generated Ballerina bindings, Ballerina primitive types, Ballerina string type, and the generated Ballerina objects will be mapped internally onto the respective Java primitives, Java String object, and the respective handle references of the objects.

#### Support for Java Subtyping
Ballerina bindings provide support for Java subtyping with the aid of its structural typing.

> E.g., A Ballerina object mapping the Java class `java.io.FileInputStream` could be assigned to a Ballerina object mapping `java.io.InputStream` as follows.

Reviewer: ```suggestion
E.g., A Ballerina object mapping the `java.io.FileInputStream` Java class could be assigned to a Ballerina object mapping the `java.io.InputStream` as follows.
```

```ballerina
InputStream inputStream = check newFileInputStream("sample.txt");
```

#### Support for Java Casting
The `ballerina/java` module of the Ballerina standard library provides the `cast` function to support Java casting. This could be used to cast Ballerina binding objects into its subtypes based on assignability.

E.g., A Ballerina object mapping the Java class `java.io.InputStream` could be casted into a Ballerina object mapping `java.io.FileInputStream` as follows.
```ballerina
FileInputStream fileInputStream = <FileInputStream>check java:cast(inputStream, typedesc<FileInputStream>);
```

## Packaging Java libraries with Ballerina programs

> This section assumes that you have already read the guide [How to Structure Ballerina Code](https://ballerina.io/v1-1/learn/how-to-structure-ballerina-code/). When you compile a Ballerina program with `ballerina build <root-module>`, the compiler creates an executable JAR file and when you compile a Ballerina module with `ballerina build -c <module>`, the compiler creates a BALO file. In both cases, the Ballerina compiler produces self-contained archives. There are situations in which you need to package JAR files with these archives. The most common example would be packing the corresponding JDBC driver.

Reviewer: ```suggestion
This section assumes that you have already read [How to Structure Ballerina Code](https://ballerina.io/learn/how-to-structure-ballerina-code/). When you compile a Ballerina program with`ballerina build <root-module>`, the compiler creates an executable JAR file and when you compile a Ballerina module with`ballerina build -c <module>`, the compiler creates a BALO file. In both cases, the Ballerina compiler produces self-contained archives. There are situations in which you need to package JAR files with these archives. The most common example would be packing the corresponding JDBC driver.
```
